src/main.rs

```diff
@@ -289,21 +296,20 @@ impl Backend {
             self.client.log_message(msg.message_type, msg.message).await;
         }
 
-        self.client
-            .publish_diagnostics(
-                uri.clone(),
-                messages
-                    .into_iter()
-                    .map(|(message, range)| Diagnostic {
+        for (uri, message, range) in messages.into_iter() {
+            self.client
+                .publish_diagnostics(
+                    uri,
```

Comment:
Can we merge diagnostics per URI so everything for each file is published at once? Otherwise, we may send diagnostics many times to the same file, which is both slow and will cause the last sent diagnostic to be overwritten each time.

Ideally, this would also handle the above added block to clear diagnostics by sending an empty set to files which we have checked but found no errors.

Reply:
Yes I was thinking about it. First I thought that there was a single diagnostic per file but I did not realized that previous diagnostics were just overwritten.

I am fixing it now.